crates/stages/src/stages/bodies.rs

```diff
@@ -123,34 +127,61 @@ impl<DB: Database, D: BodyDownloader, C: Consensus> Stage<DB> for BodyStage<D, C
                     reached_tip: false,
                 })
             };
-            let block_number = block.number;
+
             // Write block
-            let key = (block_number, block.hash()).into();
-            // Additional +1, increments tx count to allow indexing of ChangeSet that contains block
-            // reward. This can't be added to last transaction ChangeSet as it would
-            // break if block is empty.
-            let this_tx_count = first_tx_id +
-                block.body.len() as u64 +
-                if self.consensus.has_block_reward(block.number) { 1 } else { 0 };
-            tx_count_cursor.append(key, this_tx_count)?;
-            ommers_cursor.append(
-                key,
-                StoredBlockOmmers {
-                    ommers: block.ommers.into_iter().map(|header| header.unseal()).collect(),
-                },
-            )?;
+            let block_header = response.header();
+            let block_number = block_header.number;
+            let block_key = block_header.num_hash().into();
+
+            match response {
+                BlockResponse::Full(block) => {
+                    body_cursor.append(
+                        block_key,
+                        StoredBlockBody {
+                            start_tx_id: current_tx_id,
+                            tx_count: block.body.len() as u64,
+                        },
+                    )?;
+                    ommers_cursor.append(
+                        block_key,
+                        StoredBlockOmmers {
+                            ommers: block
+                                .ommers
+                                .into_iter()
+                                .map(|header| header.unseal())
+                                .collect(),
+                        },
+                    )?;
+
+                    // Write transactions
+                    for transaction in block.body {
+                        // Insert the transaction hash to number mapping
+                        db.put::<tables::TxHashNumber>(transaction.hash(), current_tx_id)?;
+                        // Append the transaction
+                        tx_cursor.append(current_tx_id, transaction)?;
+                        tx_transition_cursor.append(current_tx_id, transition_id)?;
```

Comment:
This is OK because the Block Index transition_id will always point to the last transaction OR a reward from a block. This number will always be >= any transition_id pointed to by a transaction. Hence this is correct.